### DISCUSSION TITLE
Is Arrow functionality available via SQL?

### DISCUSSION THREAD

**Discussion by rkennedy-mode** [duckdb/duckdb#3025] [Q&A]
There's a bunch of Parquet functionality that's made available via SQL ([docs](https://duckdb.org/docs/data/parquet)). Are there similar allowances for using Arrow via SQL? I've tried using `COPY … FROM …`, `SELECT * FROM 'filename.arrow'`, and `SELECT * FROM arrow_scan('filename.arrow')` without success:

```
D select * from 'benchmark_data.arrow';
Error: Catalog Error: Table with name benchmark_data.arrow does not exist!
LINE 1: select * from 'benchmark_data.arrow';
                      ^
D select * from arrow_scan('benchmark_data.arrow');
Error: Binder Error: No function matches the given name and argument types 'arrow_scan(VARCHAR)'. You might need to add explicit type casts.
        Candidate functions:
        arrow_scan(POINTER, POINTER, UBIGINT)

LINE 1: select * from arrow_scan('benchmark_data.arrow');
                      ^
D CREATE TABLE benchmark_table (id BIGINT, message VARCHAR);
D COPY benchmark_table FROM 'benchmark_data.arrow' WITH (FORMAT ARROW);
Error: Catalog Error: Copy Function with name arrow does not exist!
Did you mean "csv"?
```

The most promising attempt was using `arrow_scan`, but I'm unsure what the indicated `POINTER, POINTER, UBIGINT` arguments would be meant to accept. Curiously, I searched the code ([results](https://github.com/duckdb/duckdb/search?q=%22arrow_scan%22)) and couldn't find anything that appeared to be the declaration of the SQL function. Is it somehow loaded from somewhere else?

**Comment by hannes** [comment]
The scan function is in the core, but indeed the actual usage is I the Python and R packages. It gets a bit involved with selection and projection pushdown, its used e.g. here: https://github.com/duckdb/duckdb/blob/9552f65a28110408a8d817b396d618697023712c/tools/pythonpkg/src/pyconnection.cpp#L195 : What did you want to do there?

**Comment by rkennedy-mode** [comment]
We're looking at alternatives to CSV import and the Appender class. The Appender class is fine-grained enough to let us parse the CSV ourselves and detect cell-by-cell errors in the data as we load it into DuckDB, but it's also about 3x slower than CSV using `COPY … FROM …`, which doesn't have very detailed, granular error reporting (e.g. if the type doesn't match or if a value is out of range).

I'm guessing an Arrow-based `COPY … FROM …` would be at least as fast as CSV via `COPY … FROM …` (though likely faster due to not needing to parse the CSV and convert base10 representations of numeric values into native numeric values) while also having a measure of built-in type checking (via the Arrow schema), possibly allowing us to bypass our own error checking for even better speed.

Our service is written in Java, so SQL/JDBC represents the majority of the functional DuckDB surface area exposed to us. That rules out the Python and R surfaces. C/C++ could be made available via the JNI bindings, but I haven't figured out how to access the Arrow functionality from those languages, either.

**Reply by jonathanswenson** [reply]
@rkennedy-mode I have been working a little on the side on this one. 

The primary challenge is that the duckdb folks don't want to include the Arrow IPC libraries as a dependency. I came up with a slightly odd solution that completely avoid including the libraries in the JDBC driver that is shipped, but needs some larger structural changes to how the tests are compiled / executed / etc in order to still be able to write tests for arrow ingestion. 

You can see the primary work here: https://github.com/duckdb/duckdb/pull/4071

I also have other branches
+ where I've gone and added testing: https://github.com/jonathanswenson/duckdb/commit/77ec47eb367755e3412aa5a3f2b3fb0b354d84f9#diff-82aedb94b8a051102f611bcc51b8027756d657ba5e371439bbd22f389e6e745eR52-R98
+ and played with some of the build tools to build / test with maven rather than trying to fake it with cmake: https://github.com/jonathanswenson/duckdb/commit/60055f0448ba6cc2254de08d32154685643d60dc

While the primary PR only explores arrow ingestion, the testing / additional work also digs into what it would take to export arrow data (much more complicated from an interface perspective without including arrow as a dependency). 

However with these changes (or something like them) you could implement duckdb ingestion like so (this is kotlin): 
```
val vectorSchemaRoot = arrowReader.vectorSchemaRoot
val schema = vectorSchemaRoot.schema

val wrapper = object : DuckDBArrayStreamWrapper {
    override fun exportArrayStream(arrayStreamPointer: Long) {
        val arrowArrayStream = ArrowArrayStream.wrap(arrayStreamPointer)
        Data.exportArrayStream(allocator, arrowReader, arrowArrayStream)
    }

    override fun exportSchema(schemaPointer: Long) {
        val arrowSchema = ArrowSchema.wrap(schemaPointer)
        Data.exportSchema(allocator, schema, null, arrowSchema)
    }
}

duckDBConnection.ingestArrowTable("main", tableName, wrapper)
```

and exporting a result set like this (more kotlin): 
```
val duckResultSet = resultSet as DuckDBResultSet

val importer = DuckDBArrayStreamImporter { arrayStreamPointer ->
    val stream = ArrowArrayStream.wrap(arrayStreamPointer)
    Data.importArrayStream(allocator, stream)
}

// this cast is required due to not including the dependency within the library.
val arrowReader = duckResultSet.arrowExportArrayStream(importer) as ArrowReader

val vsr = arrowReader.vectorSchemaRoot
val field = vsr.schema.fields.first()

while (arrowReader.loadNextBatch()) {
    // TODO iterate through results in this vector
    val fieldVector = vsr.getVector(field)
    println(fieldVector.getObject(0))
}
```

**Comment by hannes** [comment]
Maybe we can expose more details from the import to make the COPY route feasible, what would you need?

**Reply by rkennedy-mode** [reply]
Our current import reports the row number, the attribute/column name and attribute/column type of any errors (or at least the first one encountered) that we find during the loading process (e.g. "row=38; column=updated_at; type=timestamp"), as well as some text explanation of what the problem was (e.g. "Not a valid date[time]" or "Value out of range for type").

From Java it gets a little difficult because all we get is a JDBC SQLException with a message. If the text is guaranteed to be structured in a particular way we can get away with a regular expression to extract this detail (it's not ideal, but doable). Alternatively, we could type check and cast to a SQLException subclass (e.g. DuckDBCopySQLException or something better named) and make some basic getters available on that class.

**Reply by hannes** [reply]
Okay this is something we are planning to add, that the CSV reader does not blindly throw an issue but you are able to inspect these issues (e.g. in a temp "load_report" table or so. The info you have mentioned is clearly doable.

**Comment by rkennedy-mode** [comment]
It might also be worth understanding the differences between `COPY … FROM …` and `arrow_scan` from a repeated query standpoint. We're looking at `COPY … FROM …` because we figure that will guarantee that the data is going to be cached in RAM (we're using an in-memory database) for subsequent queries against the same data.

But if DuckDB internally caches much/all of what it's reading via `arrow_scan` for use in followup queries then it may be functionally equivalent and we may want to consider `SELECT * FROM arrow_scan(…)` type access patterns to avoid the overhead of DDL/table management/etc.

**Reply by divyar1101** [reply]
Hi @hannes Is there a plan to add native support for Arrow in DuckDB anytime soon?

**Reply by Mytherin** [reply]
DuckDB has quite comprehensive arrow support, could you be more specific about what you want to see added?
